apps/dashboard/lib/trpc/routers/key/create.ts

```diff
@@ -115,6 +118,36 @@ export const createKey = rateLimitedProcedure(ratelimit.create)
             "We are unable to create the key. Please contact support using support.unkey.dev",
         });
       });
+    if (input.recoverEnabled && keyAuth?.storeEncryptedKeys) {
+      const vault = new Vault(env().AGENT_URL, env().AGENT_TOKEN);
+      const encryptReq: EncryptRequest = {
+        keyring: workspace?.id,
+        data: key,
+      };
+      const requestId = crypto.randomUUID();
+      const context: RequestContext = { requestId };
+      const vaultRes = await vault.encrypt(context, encryptReq).catch((_err) => {
+        throw new TRPCError({
+          code: "INTERNAL_SERVER_ERROR",
+          message: "Encryption Failed. Please contact support using support@unkey.dev",
+        });
+      });
+      await db
+        .insert(schema.encryptedKeys)
+        .values({
+          keyId: keyId,
+          workspaceId: workspace?.id,
+          encrypted: vaultRes.encrypted,
+          encryptionKeyId: vaultRes.keyId,
+        })
+        .catch((_err) => {
+          throw new TRPCError({
+            code: "INTERNAL_SERVER_ERROR",
+            message:
+              "We are unable to store encrypt the key. Please contact support using support@unkey.dev",
+          });
+        });
+    }
```

Comment:
This doesn't resolve the issue of key inserts even if the encryption fails. 

You need to handle rollback of the key creation if the encryption fails. You either need to wrap the transaction in the create transaction, or you need to handle the rollback manually. Otherwise an end user will have extra keys for no reasons.

It should be something like:

Happy path:
1. Create Key
2. Encrypt key 
3. Return success

Unsuccessful: 

1. Create Key
2. Encyption or insert fails 
3. Make sure key is removed from the DB 
4. Return error. 

Some reading.

https://orm.drizzle.team/docs/transactions#transactions